Makefile

```diff
@@ -141,33 +165,37 @@ LOG_PERF = $(LOG_FILE)_perf.data
 all: $(NAME)
 
 # ******************************
-# *     NAME linkage           *
+# *     Link target            *
 # ******************************
 
-# Linking the NAME target
-$(NAME): $(OBJS)
+# Linking targets
+# NAME
+# TEST
+# TEST_SANI
+$(NAME) $(TEST) $(TEST_SANI): $(OBJS)
 	@printf "$(YELLOW)$(BOLD)link binary$(RESET) [$(BLUE)$@$(RESET)]\n"
-	$(SILENT)$(CXX) $(OBJS) -o $@
+	$(SILENT)$(CXX) $(LDFLAGS) $(OBJS) $(LDLIBS) -o $@
 	@printf "$(YELLOW)$(BOLD)compilation successful$(RESET) [$(BLUE)$@$(RESET)]\n"
-	@printf "$(BOLD)$(GREEN)$(NAME) created!$(RESET)\n"
+	@printf "$(BOLD)$(GREEN)$@ created!$(RESET)\n"
 
 # ******************************
 # *     Special targets        *
 # ******************************
 
 # Alias for creating unittests
 .PHONY: test
+# Reconfigure flags for linking with gtest
+test: CXXFLAGS = -Wall -Werror -Wextra -g
+test: LDLIBS = -lpthread -lgtest -lgmock -lgtest_main
 test: $(TEST)
 
-# Reconfigure flags for linking with gtest
-$(TEST): CXXFLAGS = -Wall -Werror -pthread
-# Set file counter to number of test files + object files
-$(TEST): TOTAL_FILES := $(words $(TEST_OBJS))
-$(TEST): $(TEST_OBJS)
-	@printf "$(YELLOW)$(BOLD)link $(TEST)$(RESET) [$(BLUE)$@$(RESET)]\n"
-	$(SILENT)$(CXX) $(TEST_OBJS) -lgtest -lgmock -lgtest_main -o $(TEST)
-	@printf "$(YELLOW)$(BOLD)compilation successful$(RESET) [$(BLUE)$@$(RESET)]\n"
-	@printf "$(BOLD)$(GREEN)$(TEST) created!$(RESET)\n"
+# Alias for creating unittests with sanitizers enabled
+.PHONY: test_sani
+# Reconfigure flags for linking with gtest and sanitizers
```

Comment:
So we are checking the unit tests for memory leaks?

Reply:
This target compiles the unittests with Address sanitizer (ASan) and Unbdefined Behavior Sanitizer (UBSan)
- [ASan ](https://github.com/google/sanitizers/wiki/AddressSanitizer) can detect some memory leaks but its better at detecting memory errors, like use after free, stack buffer overflow etc
- [UBSan](https://clang.llvm.org/docs/UndefinedBehaviorSanitizer.html) detects undefined behavior

The sanitizers are more or less free to run.
But they don't interact well with valgrind. So to also cross check with valgrind we have two different targets.